generator/internal/predicate_utils.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "generator/internal/predicate_utils.h"
+#include "google/cloud/optional.h"
+#include <string>
+
+namespace google {
+namespace cloud {
+namespace generator_internal {
+
+using ::google::protobuf::Descriptor;
+using ::google::protobuf::FieldDescriptor;
+using ::google::protobuf::MethodDescriptor;
+
+// https://google.aip.dev/client-libraries/4233
+google::cloud::optional<std::pair<std::string, std::string>>
+DeterminePagination(google::protobuf::MethodDescriptor const* method) {
+  std::string paginated_type;
+  Descriptor const* request_message = method->input_type();
+  FieldDescriptor const* page_size =
+      request_message->FindFieldByName("page_size");
+  if (!page_size || page_size->type() != FieldDescriptor::TYPE_INT32) return {};
+  FieldDescriptor const* page_token =
+      request_message->FindFieldByName("page_token");
+  if (!page_token || page_token->type() != FieldDescriptor::TYPE_STRING)
+    return {};
+
+  Descriptor const* response_message = method->output_type();
+  FieldDescriptor const* next_page_token =
+      response_message->FindFieldByName("next_page_token");
+  if (!next_page_token ||
+      next_page_token->type() != FieldDescriptor::TYPE_STRING)
+    return {};
+
+  std::vector<std::tuple<std::string, Descriptor const*, int>>
+      repeated_message_fields;
+  std::cout << "response_message->field_count(): "
```

Comment:
Was this cout accidentally left in here? In general, we discourage printing output from libraries. My suggestion would be to remove the library output, or replace it with calls to `GCP_LOG` or something.
WDYT?